d3ploy/no_inst.py

```diff
@@ -0,0 +1,308 @@
+    )
+
+    supply_std_dev = ts.Double(
+        doc="The number of standard deviations off mean for ARMA predictions",
+        tooltip="Std Dev off mean for ARMA",
+        uilabel="Supple Std Dev",
+        default=0.
+    )
+
+    demand_std_dev = ts.Double(
+        doc="The number of standard deviations off mean for ARMA predictions",
+        tooltip="Std Dev off mean for ARMA",
+        uilabel="Demand Std Dev",
+        default=0.
+    )
+
+    steps = ts.Int(
+        doc="The number of timesteps forward for ARMA or order of the MA",
+        tooltip="Std Dev off mean for ARMA",
+        uilabel="Demand Std Dev",
+        default=1
+    )
+    back_steps = ts.Int(
+        doc="This is the number of steps backwards from the current time step" +
+            "that will be used to make the prediction. If this is set to '0'" +
+            "then the calculation will use all values in the time series.",
+        tooltip="",
+        uilabel="",
+        default=10
+    )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.commodity_supply = defaultdict(float)
+        self.commodity_demand = defaultdict(float)
+        self.fac_supply = {}
+        CALC_METHODS['ma'] = self.moving_avg
+        CALC_METHODS['arma'] = self.predict_arma
+        CALC_METHODS['arch'] = self.predict_arch
+
+    def enter_notify(self):
+        super().enter_notify()
+        lib.TIME_SERIES_LISTENERS[self.supply_commod].append(self.extract_supply)
+        lib.TIME_SERIES_LISTENERS[self.demand_commod].append(self.extract_demand)         
+
+    def tick(self):
+        """
+        This is the tock method for the institution. Here the institution determines the difference
+        in supply and demand and makes the the decision to deploy facilities or not.     
+        """
+        time = self.context.time
+        diff, supply, demand = self.calc_diff(time-1)
+        print(diff, supply, demand)
+        if  diff < 0:
+            proto = random.choice(self.prototypes)
+            ## This is still not correct. If no facilities are present at the start of the
+            ## simulation prod_rate will still return zero. More complex fix is required.
+            if self.fac_supply[proto]:
+                prod_rate = self.fac_supply[proto]
+            else:
+                print("No facility production rate available for " + proto)                
+            number = np.ceil(-1*diff/prod_rate)
+            i = 0
+            while i < number:
+                self.context.schedule_build(self, proto)
+                i += 1
+        if self.record:
+            with open(self.demand_commod+".txt", 'a') as f:
+                f.write("Time " + str(time-1) + " Deployed " + str(len(self.children)) + 
+                                              " supply " + str(self.commodity_supply[time-1]) + 
+                                              " demand " +str(self.commodity_demand[time-1]) + "\n")    
```

Comment:
non-blocking, but stylistically, I prefer to create a string, assign it to a variable, and then write out the variable.